### PR TITLE
fix: make mocha self-contained with the source map support

### PR DESCRIPTION
**Pull request by raymondfeng**
This is a follow-up to https://github.com/strongloop/loopback-next/pull/884.
The commit makes it impossible to run `npm test` for individual packages.

- There is no local `mocha` command for each package and it depends on the
  global installation of `mocha`
- With the global mocha, `--require source-map-support/register` cannot be
  resolved.

The PR adds `lb-mocha` command so that we can configure `source-map-support`
correctly along with other options. No global mocha installation is needed.

## Checklist

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] Related API Documentation was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `packages/example-*` were updated

### CODE REVIEW

**Comment by shimks** [comment]
Can we converge back `mocha.js.opts` with `mocha.ts.opts`? They contain the same content and and `source-map-support` stuff is done in `run-mocha` anyway

**Comment by raymondfeng** [comment]
@shimks I consolidated it back to `mocha.opts`.

File: packages/build/README.md [hunk 1]
```diff
@@ -36,10 +36,10 @@ To use `@loopback/build` for your package:
     "clean": "lb-clean loopback-grpc*.tgz dist*",
     "prepare": "npm run build && npm run build:apidocs",
     "pretest": "npm run build:current",
-    "acceptance": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/acceptance/**/*.js'",
-    "integration": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/integration/**/*.js'",
-    "test": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/unit/**/*.js' 'DIST/test/integration/**/*.js' 'DIST/test/acceptance/**/*.js'",
-    "unit": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/unit/**/*.js'",
+    "acceptance": "lb-mocha --opts ../../test/mocha.opts 'DIST/test/acceptance/**/*.js'",
+    "integration": "lb-mocha --opts ../../test/mocha.opts 'DIST/test/integration/**/*.js'",
+    "test": "lb-mocha --opts ../../test/mocha.opts 'DIST/test/unit/**/*.js' 'DIST/test/integration/**/*.js' 'DIST/test/acceptance/**/*.js'",
```
Comment (bajtos) [inline packages/build/README.md]: Ditto. I believe `lb-mocha` is already adding `--opts` to CLI arguments.

Please make sure these scripts actually work when executed from `packages/build` directory. I would expect to see `../bin/run-mocha` here, because it is my understanding that `node_module/.bin/lb-mocha` is created only when `@loopback/build` is installed as a (dev)dependency, which is not the case in `@loopback/build` itself.

File: packages/cli/generators/project/templates/package.json [hunk 2]
```diff
@@ -35,7 +35,7 @@
 <% } -%>
     "pretest": "npm run clean && npm run build",
 <% if (project.mocha) { -%>
-    "test": "lb-dist mocha DIST/test",
+    "test": "lb-mocha 'DIST/test'",
```
Comment (bajtos) [inline packages/cli/generators/project/templates/package.json]: AFAIK, single quotes don't work on Windows, that's why we have to use escaped double quotes in package scripts.

```json
{
  "test": "lb-mocha \"DIST/test\""
}
```

File: packages/cli/package.json [hunk 3]
```diff
@@ -54,7 +55,7 @@
   },
   "scripts": {
     "prepublishOnly": "nsp check",
-    "test": "mocha --opts node_modules/@loopback/build/mocha.js.opts \"test/*.js\""
+    "test": "lb-mocha 'test/*.js'"
```
Comment (bajtos) [inline packages/cli/package.json]: Ditto, please preserve original double quotes here and also in all other package scripts you have modified in this patch.

See https://stackoverflow.com/q/10737283/69868 and https://stackoverflow.com/a/24181667/69868

File: packages/authentication/package.json [hunk 4]
```diff
@@ -6,18 +6,18 @@
     "node": ">=6"
   },
   "scripts": {
-    "acceptance": "lb-dist mocha --opts node_modules/@loopback/build/mocha.ts.opts 'DIST/test/acceptance/**/*.js'",
+    "acceptance": "lb-mocha 'DIST/test/acceptance/**/*.js'",
```
Comment (bajtos) [inline packages/authentication/package.json]: Please change single quotes to double quotes while you are changing test-related scripts in this file. (See my other comments below for more details.)

File: packages/build/README.md (hunk 1 above)
Comment (bajtos) [reply packages/build/README.md]: Also `--opts ../../test/mocha/opts` looks suspicious to me, isn't this path pointing to monorepo-level `test` folder that no longer exists? I think we should not be needed `--opts` at all, since `lb-mocha` (`run-mocha`) is already adding them.

**Comment by bajtos** [comment]
One more comment: please update `packages/build/README.md` to reflect the changes you are making here. The current content does not look correct to me:

```
    "acceptance": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/acceptance/**/*.js'",
    "integration": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/integration/**/*.js'",
    "test": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/unit/**/*.js' 'DIST/test/integration/**/*.js' 'DIST/test/acceptance/**/*.js'",
    "unit": "lb-dist mocha --opts ../../test/mocha.opts 'DIST/test/unit/**/*.js'",
```

File: packages/build/bin/run-mocha.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+'use strict';
+
+function run(argv, dryRun) {
+  const utils = require('./utils');
+  const path = require('path');
+
+  // Substitute the DIST variable with the dist folder
+  const dist = utils.getDistribution();
+  const mochaOpts = argv.slice(2).map(a => a.replace(/\bDIST\b/g, dist));
```
Comment (bajtos) [inline packages/build/bin/run-mocha.js]: Since this has effectively inlined `select-dist` script, I think we can remove `select-dist` entirely? Just make sure to mention that removal in the commit message as a BREAKING CHANGE.

**Comment by bajtos** [comment]
@raymondfeng when you rebase on top of recently landed https://github.com/strongloop/loopback-next/pull/908, please update `package.json` in `example-log-extension` too.

**Comment by shimks** [comment]
When I run `npm t` on the monorepo, the tests in the `cli/test/sandbox/test` also seem to be running. Could you also fix that with this PR please?

**Comment by raymondfeng** [comment]
PTAL

File: packages/build/bin/run-mocha.js (hunk 5 above)
Comment (raymondfeng) [reply packages/build/bin/run-mocha.js]: We have to leave `lb-dist` there to keep backward compatibility as projects generated from previous versions of our CLI reference `lb-dist`.

**Comment by raymondfeng** [comment]
> single quotes don't work on Windows

Fixed with `\"`.

> I think lb-mocha is not available inside @loopback/build package

That's expected. That's why we only use `lb-mocha` for modules that have `@loopback/build` as a dependency.

File: packages/build/bin/run-mocha.js (hunk 5 above)
Comment (bajtos) [reply packages/build/bin/run-mocha.js]: Makes sense.

To me, this is another argument against using`.alpha.X` prereleases for so long. If we were versioning non-core modules as `0.x.y`, then we could treat semver-patch release as backwards compatible and semver-minor releases as breaking. This works great with the carrot operator as implemented by npm's semver (https://www.npmjs.com/package/semver#caret-ranges-123-025-004):

> this allows patch and minor updates for versions `1.0.0` and above, patch updates for versions `0.X >=0.1.0`, and no updates for versions `0.0.X`.

File: packages/metadata/test/unit/inspector.test.ts [hunk 6]
```diff
@@ -631,6 +633,15 @@ describe('Inspector for design time metadata', () => {
     expect(meta).to.eql(MyClass);
   });
 
+  it('inspects design time type for properties with union type', () => {
```
Comment (bajtos) [inline packages/metadata/test/unit/inspector.test.ts]: This looks like unrelated changes to me. @raymondfeng could you please check these changes were intended?

File: packages/build/bin/run-mocha.js (hunk 5 above)
Comment (raymondfeng) [reply packages/build/bin/run-mocha.js]: Agree with the versioning. Is it too late to fix it?

Comment (bajtos) [reply packages/build/bin/run-mocha.js]: > Agree with the versioning.

👍 

> Is it too late to fix it?

I don't know. Let's open a new issue and try to find out what the implications of such change would be & how much work this change would require:

https://github.com/strongloop/loopback-next/issues/954
